crates/matrix-sdk/src/test_utils/mocks.rs

```diff
@@ -0,0 +1,514 @@
+pub struct MatrixMockServer {
+    server: MockServer,
+    client: Client,
+
+    /// Make the sync response builder stateful, to keep in memory the batch
+    /// token and avoid the client ignoring subsequent responses after the first
+    /// one.
+    sync_response_builder: Arc<Mutex<SyncResponseBuilder>>,
+}
+
+impl MatrixMockServer {
+    /// Create a new `wiremock` server specialized for Matrix usage.
+    pub async fn new() -> Self {
+        let server = MockServer::start().await;
+        let client = logged_in_client(Some(server.uri().to_string())).await;
+        Self { client, server, sync_response_builder: Default::default() }
+    }
+
+    /// Creates a new [`MatrixMockServer`] when both parts have been already
+    /// created.
+    pub fn from_parts(server: MockServer, client: Client) -> Self {
+        Self { client, server, sync_response_builder: Default::default() }
+    }
+
+    /// Return the underlying client.
+    pub fn client(&self) -> Client {
+        self.client.clone()
+    }
+
+    /// Return the underlying server.
+    pub fn server(&self) -> &MockServer {
+        &self.server
+    }
+
+    /// Overrides the sync/ endpoint with knowledge that the given
+    /// invited/joined/knocked/left room exists, runs a sync and returns the
+    /// given room.
+    pub async fn sync_room(&self, room_id: &RoomId, room_data: impl Into<AnyRoomBuilder>) -> Room {
+        let any_room = room_data.into();
+
+        self.mock_sync()
+            .ok_and_run(move |builder| match any_room {
+                AnyRoomBuilder::Invited(invited) => {
+                    builder.add_invited_room(invited);
+                }
+                AnyRoomBuilder::Joined(joined) => {
+                    builder.add_joined_room(joined);
+                }
+                AnyRoomBuilder::Left(left) => {
+                    builder.add_left_room(left);
+                }
+                AnyRoomBuilder::Knocked(knocked) => {
+                    builder.add_knocked_room(knocked);
+                }
+            })
+            .await;
+
+        self.client.get_room(room_id).expect("look at me, the room is known now")
+    }
+
+    /// Overrides the sync/ endpoint with knowledge that the given room exists
+    /// in the joined state, runs a sync and returns the given room.
+    pub async fn sync_joined_room(&self, room_id: &RoomId) -> Room {
+        self.sync_room(room_id, JoinedRoomBuilder::new(room_id)).await
+    }
+
+    /// Verify that the previous mocks expected number of requests match
+    /// reality, and then cancels all active mocks.
+    pub async fn verify_and_reset(&self) {
+        self.server.verify().await;
+        self.server.reset().await;
+    }
+}
+
+// Specific mount endpoints.
+impl MatrixMockServer {
+    /// Mocks a sync endpoint.
+    pub fn mock_sync(&self) -> MockSync<'_> {
+        let mock = Mock::given(method("GET"))
+            .and(path("/_matrix/client/r0/sync"))
```

Comment:
aside: shouldn't we be using `/v3/` endpoints? `/r0/` is supported by synapse for compatibility with old applications, but new codebases shouldn't be using it.

Reply:
Yeah the default Client is configured with no information about matrix versions, so it supposes the first version by default. It usually doesn't matter, except in the few cases where it does, of course. I'll ask about this in the Rust room, but i'm tempted to keep it like this right now.

Reply:
In reality we do use the `/v3` endpoints, but in tests we don't bother filling out the supported version in the `Client` and as @bnjbvr says, it falls back to a default version.

We could modify the constructor of the test `Client` object to have the versions prefilled, but that will automatically mean that you need to update all tests that are using this test client.

I would say that we should do that, but not as part of this PR.